### ISSUE TITLE
Use statement_descriptor_suffix when making card payments

### ISSUE THREAD

**Issue by james-allan** [woocommerce/woocommerce-gateway-stripe#2830]
**Describe the bug**

According to [this Stripe doc](https://support.stripe.com/questions/use-of-the-statement-descriptor-parameter-on-paymentintents-for-card-charges), using the `statement_descriptor` parameter for PaymentIntents for card charges is no longer supported and as of 01/02/2024 (assuming an American date format - Jan 2nd 2024). We should be using `statement_descriptor_suffix` instead. 

**To Reproduce**

> [!NOTE] 
> I haven't been able to replicate the issue however we have had a number of reports from customers.  These reproduce steps are assumed based on the notice from Stripe. 

1. Enable Stripe as a payment gateway.
2. Enter valid Stripe API key credentials.
3. Enable card payments. 
4. In the Stripe Payment gateway settings under **Payments & transactions** make sure there's a statement descriptor added.
5. Attempt to make a card payment from the front end.
6. According to the Stripe docs and from user reports. This should result in a failure. 

**Expected behavior**

We should switch to `statement_descriptor_suffix` for card payments specifically to avoid these failures.

**Comment by AashikP** [comment]
User Reports:

* 7573985-zen
* 7570774-zen
* 7541672-zen
* https://wordpress.org/support/topic/statement_descriptor-error/
* 7674291-zen
* 7656468-zen

**Comment by anastas10s-afk** [comment]
7582708-zen

**Comment by james-allan** [comment]
I'm having a hard time wrapping my head around how we resolve this issue the right way and honour the current Stripe plugin settings.

From my reading of Stripe's guide on [statement descriptors](https://stripe.com/docs/payments/account/statement-descriptors), the suffix (this new parameter we should be using), is automatically added to the short statement description that is stored in the Stripe Account settings ([here](https://dashboard.stripe.com/settings/public)).

This one:

<p align="center">
<img width="353" alt="Screenshot 2024-01-15 at 8 32 15 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/7feb4bd5-e0dc-48c7-be72-2ec19f8365f4">
</p> 


Previously we would allow merchants to specify the short description in our Stripe plugin settings, and we'd build the full statement description on the fly for each payment. See [`WC_Stripe_Helper::get_dynamic_statement_descriptor()`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/7.9.0/includes/class-wc-stripe-helper.php#L544-L561). 

What we were doing before no longer seems possible because we can't pass a prefix and a suffix. The one that is stored on the account level is used no matter what.


So I think the best we can hope for is something like this:

<p>
<img width="500" alt="Screenshot 2024-01-15 at 8 32 15 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/b0f0a429-4e76-4721-bce8-dcfb2ab6b1eb">
</p> 

Whereas, before it was this: 


<p>
<img width="330" alt="Screenshot 2024-01-15 at 8 32 15 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/6674a53c-68fb-43fc-aba3-f40d79ea4a21">
</p> 

This looks would be fine, however, the total length of this string including the static part saved in Stripe settings is 22 characters and we previously allowed merchants to store descriptions as long as 10 characters assuming that the order ID + the # + spaces would leave ample space. 

Anyway, I'm not sure what we want to do here. Either get rid of plugin setting and tell the merchant to set the short description in the stripe settings, or if we just move forward with the approach I mentioned above. If we do that though we'll need to work out how we trim the string down if its combined length is too long.

**Comment by morleydigital** [comment]
Just to add to this, having the same issue but only on 2 out of 3 live sites. Looking at the stripe logs, the site that is still working shows that the request used API Version 2019-09-09 (even though the default API on the account was set to 2020-08-27), whereas the two sites that are not working are using API Version 2023-10-16. The two new sites are using more recently created Stripe accounts and the default API version on those accounts is 2023-10-16.

Not sure if this helps or not, but may explain why you haven't bene able to recreate the issue.

**Comment by diegocurbelo** [comment]
> Anyway, I'm not sure what we want to do here. Either get rid of plugin setting and tell the merchant to set the short description in the stripe settings, or if we just move forward with the approach I mentioned above. If we do that though we'll need to work out how we trim the string down if its combined length is too long.

I think that if we can't use our own descriptors, and we can't update Stripe's using an API call... then, the best solution would be to replace the inputs in the plugin settings with a message saying that the Statement Descriptions can (must?) be changed on the Stripe Dashboard and include a link to the dashboard: https://dashboard.stripe.com/settings/public

Keeping the option to add the order number as a prefix, using the `statement_descriptor_suffix` attribute: 
![Screenshot 2024-01-15 at 20 41 19](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/407542/78b7051d-2e5c-40fc-b788-d708b67c4a7b)

**Comment by rossviviano** [comment]
7598200-zen

**Comment by fernmed** [comment]
7606730-zen

**Comment by NathalyaVides** [comment]
7628700-zen

**Comment by kgeorgee** [comment]
7640450-zen

**Comment by RiaanKnoetze** [comment]
7634285-zen

**Comment by rossviviano** [comment]
7634464-zen

**Comment by afzalpansuvi** [comment]
7661681-zen

**Comment by shameemreza** [comment]
Another report here: 7663469-zen
Thread link: https://wordpress.org/support/topic/can-not-pay-by-stripe/

**Comment by sverleis** [comment]
Reported on 7674291-zen and 7656468-zen

**Comment by petertwise** [comment]
Quick fix via WP CLI if you need it:

`wp option patch update woocommerce_stripe_settings statement_descriptor ''`

But yeah - this is clearly a widespread and urgent issue.

**Comment by theabhig** [comment]
Another in 7691155-zen

**Comment by julinedossantos** [comment]
Hey @james-allan, I see that you seem to have fixed this issue with the pull request: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2833

Can you confirm which version of WooCommerce-Gateway-Stripe will include it?

**Comment by diegocurbelo** [comment]
Hey @julinedossantos 👋🏻
We are releasing version 7.9.2 tomorrow with this fix.

**Comment by shameemreza** [comment]
Another report here: 7702148-zen

**Comment by shameemreza** [comment]
It seems like the issue is still ongoing on the latest version: 7828424-zen
Thread: https://wordpress.org/support/topic/stripe-descriptor-error

**Comment by solstudioim** [comment]
Affected: 7586378-zen with the latest version

**Comment by rogerArtefacto** [comment]
ar there any news on how to solve it? Even if it is using an extra function ... Haven't been able to solve it, thinking about using a different plugin

**Comment by sverleis** [comment]
> ar there any news on how to solve it? Even if it is using an extra function ... Haven't been able to solve it, thinking about using a different plugin

@rogerArtefacto , for the time being, as a workaround, are you happy to remove the description from the Customer bank statement area?
**WooCommerce > Settings > Payments > Stripe > Settings**

**Comment by rogerArtefacto** [comment]
It is taking it from my Stripe account, can't edit it on woocommerce...and cant leave it blank on Stripe...

edit: I even tried to remove it from the SQL options table, but It does not work

**Comment by shameemreza** [comment]
Another report with the latest version: 7851393
Forum thread: https://wordpress.org/support/topic/can-not-pay-by-stripe-statement_descriptor-error/

**Comment by diegocurbelo** [comment]
Hey @rogerArtefacto!.

> ar there any news on how to solve it? Even if it is using an extra function
 
There are a few edge cases where the statement descriptor does not behave as expected; we are currently working on a fix for this.

In the meantime, it seems that a snippet like this would fix your problem:
```
add_filter( 'wc_stripe_generate_create_intent_request', function ( $request, $order, $prepared_source ) {
	$request[ 'statement_descriptor_suffix' ] = WC_Stripe_Helper::get_dynamic_statement_descriptor_suffix( $order );
	unset( $request[ 'statement_descriptor' ] );
	return $request;
}, 10, 3 );
```

**Comment by rogerArtefacto** [comment]
already tried it...and did not work ... had to move to a different plugin this time (will check this one again for my next project, since I love its simplicity, but couldnt afford to tell the client he'd have no payment platform)

**Comment by a-danae** [comment]
Hey there 👋 

We have [this alpha version for 8.0.1](https://github.com/woocommerce/woocommerce-gateway-stripe/releases/tag/8.0.1-alpha) that should fix this issue. 

Please feel free to give it a go. The changes should be safe at this point but we recommend using a staging site for this since it's an alpha version.

**Comment by why13ee** [comment]
I had this problem. Talked to Stripe support, and they refer me to this page: https://docs.stripe.com/api/payment_intents/create#create_payment_intent-statement_descriptor_suffix

I'll just put it here for record.

I tried the [Alpha version 8.0.1](https://github.com/woocommerce/woocommerce-gateway-stripe/releases/tag/8.0.1-alpha). It worked. Thank you.

**Comment by a-danae** [comment]
Hey @why13ee !

Thanks for letting us know that the 8.0.1-alpha worked!

We've just released 8.0.1 with the contents of this alpha.

**Comment by CodeConnectPtyLtd** [comment]
The ability to set only the statement_descriptor_suffix in the updated plugin is still resulting chargebacks and disputes from our customers.

We have legal entity that has several trading names.  Before these changes we were able to set that trading name per trading  WooCommerce store.

Now we can only tack on the order number which also makes the static statement description basically unreadable.

Stripe's docs state that for card charges a dynamic statement descriptor is possible if the statement_descriptor_suffix is also included.

Stripe Docs:
https://docs.stripe.com/get-started/account/statement-descriptors#dynamic

> For card charges, providing a dynamic statement descriptor requires the statement_descriptor_suffix value. For non-card charges, if you set a value only for statement_descriptor on a PaymentIntent, Stripe uses it in place of the account statement descriptor (static descriptor).
